### PR TITLE
feat: implement custom error types for DA

### LINKED ISSUE #65: [Feature Request]: standard error codes
### Implementation ideas

Implementations should return standard error codes so that clients have a way to handle the error and retry if possible.

We can reuse JSONRPC error codes in the range `--32000 to -32099 `, clients should map their errors into the standard error codes for e.g.:

* 32001 - request timeout
* 32002 - mempool congested
* 32003 - insufficient fee
* 32004 - incorrect sequence / nonce
* 32005 - blob size too large

gRPC can reuse the JSONRPC error codes for simplicity.

Based on the error, the client can retry with a higher fee, different nonce or smaller blob.

Implementations should wrap their error with the corresponding typed error.

E.g.:

```go
pkg someda

...
var (
	// Err32003 is returned when the blob transaction has insufficient fee
	Err32003 = errors.New("32003")
)
...


// Submit submits the Blobs to Data Availability layer.
func (c *SomeDA) Submit(ctx context.Context, blobs []da.Blob, gasPrice float64, ns da.Namespace) ([]da.ID, error) {
        ...
	height, err := c.Submit(ctx, blobs, gasPrice, ns)
	if err != nil {
		if errors.Is(err, ErrInsufficientFee) {
			return errors.Wrap(da.Err32003, err)
		}
		return nil, err
	}
        ...
}

```

### PR DESCRIPTION
**Pull request by tzdybal**
## Overview

* replaced sentinel error with types
* defined common error codes
* added error mapping for JSON-RPC - codes are used on the transport layer
* refactored `ErrTooHigh` into `ErrFutureHeight` (renamed, moved, added error code)
* added GRPCStatus handling to error types - gRPC well-known codes are used and da-codes are passed as detail.
* fixed error handling in tests

Resolves #65 

## Summary by CodeRabbit

- **New Features**
	- Introduced structured error handling for JSON-RPC and gRPC protocols, enhancing clarity in error reporting.
	- Added new error codes and types for better management of blob and transaction errors.
	- Enhanced client and server error handling with mappings for known errors.

- **Bug Fixes**
	- Updated error handling in the implementation to use specific error types instead of generic messages.

- **Tests**
	- Improved test assertions to check for specific error types, enhancing robustness in error handling tests.

### CODE REVIEW

**Comment by tzdybal** [comment]
I'll try to define protobuf for error (code/enum) and use `WithDetails` to avoid matching by strings.

File: errors.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+// GRPCStatus returns the gRPC status with details for an ErrContextDeadline error.
+func (e *ErrContextDeadline) GRPCStatus() *status.Status {
+	return getGRPCStatus(e, codes.DeadlineExceeded, pbda.ErrorCode_ContextDeadline)
+}
+
+// GRPCStatus returns the gRPC status with details for an ErrFutureHeight error.
+func (e *ErrFutureHeight) GRPCStatus() *status.Status {
+	return getGRPCStatus(e, codes.OutOfRange, pbda.ErrorCode_FutureHeight)
+}
```
[the remark this answers is not in the document]
Comment (tzdybal) [reply errors.go]: Those methods are static - they are not depending on any values. This is very important, as for every error type the code and message is not changing. Having this as function, we don't require initialization of structs to be used.

File: proto/da/da.proto [hunk 2]
```diff
@@ -130,3 +130,19 @@ message ValidateRequest {
 message ValidateResponse {
 	repeated bool results = 1;
 }
+
+enum ErrorCode {
+	ERROR_CODE_UNSPECIFIED = 0;
```
[the remark this answers is not in the document]
Comment (tzdybal) [reply proto/da/da.proto:135]: It's not used. It's mostly because of linter complaining that first value of enum has to be `0`.

File: errors.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+)
+
+// Code defines error codes for JSON-RPC.
+//
+// They are reused for gRPC
+type Code int
+
+// Codes are used by JSON-RPC client and server
+const (
+	CodeBlobNotFound               Code = 32001
```
Comment (MSevey) [inline errors.go:17]: just out of curiosity, any specific reasoning behind using 32000s?

File: errors.go [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+	base := status.New(grpcCode, err.Error())
+	detailed, err := base.WithDetails(&pbda.ErrorDetails{Code: daCode})
+	if err != nil {
+		return base
+	}
+	return detailed
+}
+
+// GRPCStatus returns the gRPC status with details for an ErrBlobNotFound error.
+func (e *ErrBlobNotFound) GRPCStatus() *status.Status {
```
Comment (MSevey) [inline errors.go:96]: [non-blocking] nit: I'd move these into the grouping for each type for consistency. 
```
type 
func Error()
func GRPCStatus()
```

File: errors.go
Comment (MSevey) [inline errors.go:1]: [non-blocking] I think we could do the following to deduplicate some code couldn't we?

```go
type Code int

// Define all error codes
const (
   codes....
)

// Define generic type that satisfies the Errors interface
type DAError struct {
   message
   errorCode
   grpcCode
}

func (e *DAError) Error() string {
  return e.message
}

func (e *DAError) GRPCStatus() *status.Status {
    return getGRPCStatus(e, e.errorCode, e.protoCode)
}

// Define all errors
var (
    ErrBlobNotFound = &DARError{
        message:   "blob: not found",
        errorCode:  codes.NotFound,
        protoCode: pbda.ErrorCode_ERROR_CODE_BLOB_NOT_FOUND,
    }
   ...
)
```

**Review by MSevey — APPROVED** [review]
lgtm, just some thoughts that can either be addressed, resolved, or turned into good first issues.

File: errors.go (hunk 3 above)
Comment (tuxcanfly) [reply errors.go:17]: It satisfies the jsonrpc error spec: https://www.jsonrpc.org/specification#error_object

**Review by tuxcanfly — APPROVED** [review]
OK.

File: errors.go
Comment (tzdybal) [reply errors.go:1]: Yes, but I didn't want to put gRPC specific inside error itself.

**Review by MSevey — APPROVED** [review]

**Review by gupadhyaya — APPROVED** [review]

**Pull request merged**